syntax = "proto3";

service MlModule {
  rpc GetModels (GetModelsRequest) returns (GetModelsResponse);
  rpc GetHyperParams (GetHyperParamsRequest) returns (GetHyperParamsResponse);
  rpc PredictOrCreateModel (PredictOrCreateModelRequest) returns (PredictOrCreateModelResponse);
  rpc RemoveModel (RemoveModelRequest) returns (RemoveModelResponse);
}

enum TaskType {
  CLASSIFICATION = 0;
  REGRESSION = 1;
}

message GetModelsRequest {
  TaskType task_type = 1;
}

message GetModelsResponse {
  repeated string model_names = 1;
}

message GetHyperParamsRequest {
  string model_name = 1;
}

message GetHyperParamsResponse {
  repeated string hyper_params = 1;
}

message PredictOrCreateModelRequest {
  string model_name = 1;
  string hyper_params = 2;
  bool retrain = 3;
  string csv_path = 4;
  string target = 5;
}

message PredictOrCreateModelResponse {
  string status = 1;
  repeated float predict = 2;
}

message RemoveModelRequest {
  string model_name = 1;
}

message RemoveModelResponse {
  string status = 1;
}